### PR TITLE
Replace `floor()` with `round()` in pixel snap to transform

### LINKED ISSUE #56793: Vulkan: 2D Pixel Snap not working
### Godot version

4.0dev (ce6de71)

### System information

Windows 10, NVIDIA GeForce GTX 960

### Issue description

I've seen other issues like this but they're all closed as fixed. I'm still having this issue. Pixels are not snapping correctly with either or both pixel snap options.
1 pixel gap in between collisions.

### Steps to reproduce

Create a project with a floor and characterbody2d with basic movement functions, gravity, etc.
Set display stretch mode to `viewport` and `keep`.
Enable `Snap 2d Transforms to Pixel` or `Snap 2d Vertices to Pixel` or both.

image

### Minimal reproduction project

[PixelSnapBug.zip](https://github.com/godotengine/godot/files/7872520/PixelSnapBug.zip)

### LINKED ISSUE #84632: Sprite "crunch" when placed in certain positions on screen
### Godot version

v4.2.beta5.official [4c96e9676]

### System information

Godot v4.2.beta5 - Windows 10.0.22621 - Vulkan (Forward+) - dedicated AMD Radeon RX 7900 XTX (Advanced Micro Devices, Inc.; 31.0.22023.1014) - AMD Ryzen 9 7950X3D 16-Core Processor (32 Threads)

### Issue description

In 4.2, there is an issue mostly visible on pixel art games with small resolutions that scale up to larger screens. While the game is running, moving sprites to certain places in the screen will cause the sprites to appear uneven or "crunched", and it's usually more noticeable if the sprite is animated.

example

Window settings:
image

This seems to happen when moving the position of the sprite to X.5.
This happens in all 3 renderers.
This issue does not occur in Godot 4.1.3.
This issue occurs as early as 4.2 dev1.

Note: I'm using a 4K monitor with Windows scaled up to 175%.

### Steps to reproduce

1. Download the reproduction project
2. Open the project
3. Run the game
4. If the issue doesn't immediately show itself, slightly moving the character to the left/right a few times should trigger it.

### Minimal reproduction project

[SpriteCrunchIssue.zip](https://github.com/godotengine/godot/files/13301032/SpriteCrunchIssue.zip)

### PR DESCRIPTION
**Pull request by KeyboardDanni**
Fixes https://github.com/godotengine/godot/issues/56793.
Fixes https://github.com/godotengine/godot/issues/84632.

Similar to https://github.com/godotengine/godot/pull/76277 except this PR also makes the necessary changes to `renderer_viewport.cpp` to avoid regressions in camera jitter. In fact, with these changes, the camera is actually *more* stable according to my testing.

Before:

https://github.com/godotengine/godot/assets/34800072/dd708c82-9c80-4f2f-8843-7ba8ef04d0e6

After:

https://github.com/godotengine/godot/assets/34800072/8ea44763-9553-4040-90a3-1f3eb2cbe606

**NOTE:** You will still get camera jitter if you are using a script to make a camera follow an object that is moved in idle process. Camera movement is fine if the object is moved in physics process. This is an existing issue which will be addressed in a separate PR.

Test project download:

[PixelPerfectTest.zip](https://github.com/godotengine/godot/files/13244300/PixelPerfectTest.zip)

_Production edit: added fixes https://github.com/godotengine/godot/issues/84632_

### CODE REVIEW

**Comment by lawnjelly** [comment]
Related #76277 , and https://github.com/godotengine/godot/issues/57221#issuecomment-1022917087 .

**Comment by markdibarry** [comment]
Looks good! Results seem much more consistent. If there are issues I haven't found, it still seems the current implementation isn't usable as is, so this is still a big improvement.

**Comment by Cerno-b** [comment]
General thought/question: When transitioning from float to pixel coordinates, is Godot using a pixel-centered or top-left coordinates in general?

https://user-images.githubusercontent.com/16470072/162586902-f1e9ce5f-2cb7-4fab-b349-dd411eedb56f.png

This PR seems to assume pixel-centered (rounding instead of flooring). 

I am asking because I haven't found any specs on this in the docs and I think it should be defined somewhere for consistent behavior. All sorts of problems can occur if it's sometimes centered and sometimes top-left. 

This question does not block the PR, but it may be worthwhile to check if there are other places where float to pixel coordinate transition occurs and change them accordingly.

**Comment by KeyboardDanni** [comment]
My understanding is that it's using top-left for pixel coordinates, but this PR just changes when rounding occurs. That is, an object at x = 1.8 gets rounded to 2.0 instead of 1.0. Objects still snap to 1.0 or 2.0; they don't snap to 1.5, for example, with this change. With this rounding you'll want to place your objects at whole pixel coordinates for best results instead of offsetting their position by half a pixel.

Of course I imagine in practice the pixel offset might not actually be half a screen pixel and this would come down to how the pixel center is handled in either the vertex shader, 3D API, or drivers. That's an area I'm less familiar with.

**Comment by carlOSx64** [comment]
I did a custom build using your fork and it fixed all the issues I've had with pixel perfect rendering. The options right now in Godot are not working properly and you're better off using custom code in `_process` functions. I hope this gets merged soon.

**Comment by Cerno-b** [comment]
@KeyboardDanni I experimented a bit with Godot and I agree that it's top-left (corner pixel's top left is [0.0, 0.0] and bottom right is [0.99, 0.99]). Under that regimen, flooring *should* be correct to get from float it int. 

From what I've read so far, that's the way OpenGL, Direct3D 10 and Vulcan does it as well.

But don't let this hold up the PR: If rounding works here, I think it's pragmatic to merge this if it solves the current issues. 

We can always go deeper into the whys and hows with a separate proposal. I noticed another weird inconsistency in float to int, so looking at this from first principles is probably a good idea, but outside of the scope of this PR.

**Comment by KeyboardDanni** [comment]
> @KeyboardDanni I experimented a bit with Godot and I agree that it's top-left (corner pixel's top left is [0.0, 0.0] and bottom right is [0.99, 0.99]). Under that regimen, flooring _should_ be correct to get from float it int.

I think you might be confusing two different concepts here. There's the rounding behavior, i.e. the criteria that determines where to round up or down. Then there's the offset that we're rounding to, i.e. whether we round to top-left or center pixel. This PR focuses on the former.

Without this PR, the current rounding behavior means that graphics snapped via top-left pixel are going to be put right up next to the rounding boundary. This means if you have an object at position 1.0, the moment it moves to 0.999 it'll suddenly snap to 0.0. I theorize that this floor snapping, combined with floating point precision errors, is responsible for the shakiness you see in the first video.

With this PR, an object at position 1.0 needs to travel at least half a pixel before it gets rounded away from 1.0, increasing stability. Visualized:
RoundPixelSnapping

**Comment by Cerno-b** [comment]
@KeyboardDanni Thanks for the clarification. I think I may have lost myself a bit in the different kinds of coordinate systems:

https://www.realtimerendering.com/blog/the-center-of-the-pixel-is-0-50-5/
https://learn.microsoft.com/en-us/windows/win32/direct3d10/d3d10-graphics-programming-guide-resources-coordinates

From that I was under the assumption that you could either put your pixel centers at (0.5, 0.5) which would always require flooring or you could put your pixel centers at (0.0, 0.0) which would always require rounding. But this is probably a different concept from the problem that you are solving and I may have mixed up the two.

With the image examples, I see how flooring causes problems and how your rounding solves it.

**Comment by KeyboardDanni** [comment]
> you could either put your pixel centers at (0.5, 0.5) which would always require flooring or you could put your pixel centers at (0.0, 0.0) which would always require rounding.

This sounds about right, yeah.

**Review by adamscott — CHANGES_REQUESTED** [review]
Absolutely great! It seems a lot nicer with your PR, indeed.

But I tested it with @HybridEidolon's [PixelPerfect.zip](https://github.com/godotengine/godot-proposals/files/10947477/PixelPerfect.zip) (from https://github.com/godotengine/godot-proposals/issues/6389#issuecomment-1464811171) and the player's sprite falls through the ground with your PR.

I'll investigate on why changing the offset causes this side-effect.

**Review by adamscott — APPROVED** [review]
Nevermind, I rebased your PR and it seems to work with the [PixelPerfect.zip](https://github.com/godotengine/godot-proposals/files/10947477/PixelPerfect.zip) project mentionned before. 

@KeyboardDanni Can you rebase your PR?

**Comment by adamscott** [comment]
Your PR seems to fix [Rubonnek](https://github.com/Rubonnek)'s [PixelPerfectJitter.zip](https://github.com/godotengine/godot-proposals/files/13824233/PixelPerfectJitter.zip) (from https://github.com/godotengine/godot-proposals/issues/6389#issuecomment-1876131069) too!!

**Comment by adamscott** [comment]
I'm wondering though if:
 - it could break existing projects;
 - it would be better to create a project setting for the pixel snapping method.

**Comment by adamscott** [comment]
@clayjohn I closed #86670 because this PR seems better suited. It already deals with the snapping options, without having to resort to round "low-level" in the driver.

**Comment by markdibarry** [comment]
@adamscott 

> I'm wondering though if it could break existing projects

I think the risk level is low since the code changes only affect blocks where it's been established that pixel snap is turned on.

> it would be better to create a project setting for the pixel snapping method.

I believe it was decided that this is a better solution, for various reasons mentioned here: https://github.com/godotengine/godot-proposals/issues/8318

**Comment by kleonc** [comment]
Looking at this PR I've realized a big issue with the current 2D transform snapping for Sprite2D and alike: it's... applied twice. :man_facepalming:
- Firstly, it's applied when calculating the destination rect for the draw texture rect command.
- Secondly, it's applied to the Sprite2D's transform within the rendering server (transform snapping is also applied to the transform of each parent CanvasItem in the chain but it's the double-snapping for the Sprite2D itself which is absolutely wrong).

Meaning an already snapped draw texture rect command is being drawn according to a snapped Sprite2D's transform. Given that before this PR these were flooring it can very easily result in big discrepancies.

Simple example, a Sprite2D with 1x1 texture, v4.2.1.stable.official [b09f793f5] (so snapping by flooring):
| `snap_2d_transforms_to_pixel=false` | `snap_2d_transforms_to_pixel=true`<br>(by flooring) |
|--------|--------|
|0KPZSRgEdH|Godot_v4 2 1-stable_win64_MWype7FM7i|

Kinda surprising where it ended up being rendered, huh? :upside_down_face: Nearly 2 pixels away on each axis.

---

Is this still an issue if snapping is done with rounding instead of flooring? Yes, it is. The difference is that for such single double-snapping:
- by flooring `error < 2.0`,
- by rounding `error <= 1.0`.

(by error I mean per axis distance to the original unsnapped rect)

Same 1x1 texture, but with this PR (so snapping by rounding):
| `snap_2d_transforms_to_pixel=false` | `snap_2d_transforms_to_pixel=true`<br>(by rounding) |
|--------|--------|
|75KAKDBOmR|v4WTVdxhNi|

Nearly 1 pixel away, yikes!

---

Regardless, just changing flooring to rounding seems like an improvement / good first step. So it's probably fine to merge this as is, and fix the mentioned double-snapping in a subsequent PR (unless the author wants to tackle it at once in here; either one is fine for me :slightly_smiling_face:).


> I'm wondering though if:
> 
> * it could break existing projects;

Let's keep it in 4.3+ (no backports) and I guess it should be fine? Still, it's more like a bugfix/improvement and I think potential breaking by bugfixing is expectable/acceptable. :thinking: Just because users found a workaround for an issue (and use it in their projects) doesn't mean the issue shouldn't be tackled/fixed.

**Comment by KeyboardDanni** [comment]
If I had to guess, the additional snapping done in the Sprite2D code is probably meant to address the corner case of having a sprite with an odd-number size that is centered on the Node2D position in space. This needs additional investigation, though.

**Comment by markdibarry** [comment]
@KeyboardDanni In case you missed it in the last few comments. Could you rebase your PR?

**Comment by adamscott** [comment]
I tested out your PR against [Scalazard](https://github.com/nicolasbize/Scalazard), an open-source pixel perfect game.

It seems to fix a weird issue where, when pushing a box, the player and the box would jitter a bit.

### Without the PR (floor)
https://github.com/godotengine/godot/assets/270928/57124d2d-24e4-4782-b35e-079ab27e9906

### With the PR (round)
https://github.com/godotengine/godot/assets/270928/6654b863-24e6-430e-8b05-d703896a1c27

**Comment by nicolasbize** [comment]
@adamscott that's really cool to see :)

**Comment by adamscott** [comment]
I would suggest you all to checkout #87058. It seems that a fix was introduced in 3.x that was not forward-ported to 4.x.

It would remove the need of this PR as it removes the last transformations in the DisplayServer that did break everything.

Could you test your games/projects with that PR? Thanks!

**Comment by Proggle** [comment]
> I would suggest you all to checkout #87058. It seems that a fix was introduced in 3.x that was not forward-ported to 4.x.
> 
> It supersedes this PR, as we don't need to round values as it removes the last transformations in the DisplayServer that did break everything.
> 
> Could you test your games/projects with that PR? Thanks!

The issue I'm having with pixel distortion on tiles still seems unaffected, sadly.
CameraDistortExample
(gif stabilized by 1 pixel of camera movement just to make the distortion a bit more obvious)

**Comment by akien-mga** [comment]
Superseded by #87297. Thanks for the contribution!

**Pull request closed without merging**
